### PR TITLE
HTTP2 and gzip

### PR DESCRIPTION
**Pull request by Joannis**
This PR will focus on HTTP2 and Gzip.

The reason this is bundled is because HTTP2 relies on Huffman trees, which are are one of the primary components of Gzip. Gzip will not implement a good algorithm such as LZ77 yet but will instead use a primitive and unperformant implementation as a placeholder for a future PR.

Current state:

- [x] HPACK
- [x] ALPN for TLS
- [x] A _very_ basic HTTP/2 client
- [x] Body streams (with an HTTP/1.1 implementation)

More todos will be added once they're almost up for development...

### CODE REVIEW

**Comment by Joannis** [comment]
Follow up PRs after this one:

1. Server-side
2. Push notifications
3. Some general house holding/API improvements

File: Sources/HTTP/Serializer/ResponseSerializer.swift [hunk 1]
```diff
@@ -22,8 +22,7 @@ public final class ResponseSerializer: Serializer {
     /// Handles incoming responses.
     public func inputStream(_ input: Response) {
         let data = serialize(input)
-        let message = SerializedMessage(message: data, onUpgrade: input.onUpgrade)
-        outputStream?(message)
+        outputStream?(data)
```
Comment (tanner0101) [inline Sources/HTTP/Serializer/ResponseSerializer.swift]: do we not support upgrading to websockets anymore?

File: Sources/HTTP2/Error.swift [hunk 2]
```diff
@@ -0,0 +1,63 @@
+import Bits
+import Foundation
+
+struct Error: Swift.Error {
```
Comment (tanner0101) [inline Sources/HTTP2/Error.swift]: HTTP2Error

File: Sources/HTTP/Serializer/ResponseSerializer.swift (hunk 1 above)
Comment (Joannis) [reply Sources/HTTP/Serializer/ResponseSerializer.swift]: We never made use of this.

File: Sources/Async/Streams/Stream.swift [hunk 3]
```diff
@@ -0,0 +1,80 @@
+/// A stream is both an InputStream and an OutputStream
+public typealias Stream = InputStream & OutputStream
```
Comment (tanner0101) [inline Sources/Async/Streams/Stream.swift]: async module should be removed.

File: Sources/HTTP/Message/BodyStream.swift [hunk 4]
```diff
@@ -0,0 +1,31 @@
+import Async
+import Bits
+
+public final class BodyStream: Async.Stream, ClosableStream {
```
Comment (tanner0101) [inline Sources/HTTP/Message/BodyStream.swift]: comments

File: Sources/HTTP/Serializer/ChunkEncoder.swift [hunk 5]
```diff
@@ -0,0 +1,36 @@
+import Async
+import Bits
+import Foundation
+
+final class ChunkEncoder: Async.Stream, ClosableStream {
```
Comment (tanner0101) [inline Sources/HTTP/Serializer/ChunkEncoder.swift]: comments

File: Sources/HTTP/Serializer/ChunkEncoder.swift [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    
+    func close() {
+        stream.close()
+    }
+    
+    func onClose(_ onClose: ClosableStream) {
+        stream.onClose(onClose)
+    }
+    
+    /// - TODO: Improve performance
```
Comment (tanner0101) [inline Sources/HTTP/Serializer/ChunkEncoder.swift]: we should make that a private comment

File: Sources/HTTP2/Client/Client+ConnectionManagement.swift [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+            // On successful connection, send the preface
+            Constants.staticPreface.withUnsafeBytes { (pointer: BytesPointer) in
+                let buffer = ByteBuffer(start: pointer, count: Constants.staticPreface.count)
+                
+                tlsClient.onInput(buffer)
+            }
+            
+            // Send the settings, next
+            client.updateSettings(to: settings)
+            }.catch(client.promise.fail)
```
Comment (tanner0101) [inline Sources/HTTP2/Client/Client+ConnectionManagement.swift]: spacing

File: Sources/HTTP2/Client/Client+FrameInput.swift [hunk 8]
```diff
@@ -0,0 +1,65 @@
+extension HTTP2Client {
+    /// Updates the frame's contents as settings, sends an acknowledgement
+    ///
+    /// TODO: Settings changed in the middle of the stream affecting the context
```
Comment (tanner0101) [inline Sources/HTTP2/Client/Client+FrameInput.swift]: we should have these as `//` so they don't show up in the documentation

File: Tests/HTTP2Tests/HTTP2Tests.swift [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+@testable import HTTP2
+import Async
+
+public class HTTP2Tests: XCTestCase {
+    static let allTests = [
+        ("testClient", testClient)
+    ]
+    
+    func testClient() throws {
+//        #if os(Linux)
```
Comment (tanner0101) [inline Tests/HTTP2Tests/HTTP2Tests.swift]: can we get these working before merging?

File: Tests/HTTPTests/ParserTests.swift [hunk 10]
```diff
@@ -70,15 +71,52 @@ class ParserTests : XCTestCase {
...
+        Connection: Keep-Alive\r
+        \r
+        hello
+        """.data(using: .utf8) ?? Data()
+        
+        let parser = RequestParser(on: DispatchQueue.global(), maxSize: data.count - 2)
+        XCTAssertThrowsError(try parser.parse(from: data))
+    }
+    
+    func testTooLargeResponse() throws {
```
Comment (tanner0101) [inline Tests/HTTPTests/ParserTests.swift]: 👍

File: Tests/HTTP2Tests/HTTP2Tests.swift (hunk 9 above)
Comment (Joannis) [reply Tests/HTTP2Tests/HTTP2Tests.swift]: I don't have an HTTP/2 host that I can test again. External hosts like google.com can not be connected to from CircleCI

File: Sources/HTTP2/Client/Client+FrameInput.swift (hunk 8 above)
Comment (Joannis) [reply Sources/HTTP2/Client/Client+FrameInput.swift]: They won't, since they're internal APIs

**Pull request merged**